pkg/test/e2e/nodeport-proxy/services.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeportproxy
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Based on:
+// https://github.com/mgdevstack/kubernetes/blob/9eced040142454a20255ae323279a38dc6b2bc1a/test/e2e/framework/service/jig.go#L60
+// ServiceJig is a jig to help service testing.
+type ServiceJig struct {
+	Log       *zap.SugaredLogger
+	Client    ctrlclient.Client
+	Namespace string
+
+	Services    []*corev1.Service
+	ServicePods map[string][]string
+}
+
+// CreateNodePortService deploys a service and the associated pods.
+func (n *ServiceJig) CreateNodePortService(name string, nodePort int32, numPods int32, annotations map[string]string) (*corev1.Service, error) {
+	// Create the namespace to host the service
+	ns := n.newNamespaceTemplate()
+	n.Log.Debugw("Creating namespace", "service", ns)
+	if err := n.Client.Create(context.TODO(), ns); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return nil, err
+		}
+	} else {
+		// Set back namespace name in case it was generated
+		n.Namespace = ns.Name
+		n.Log.Debugw("Setting generated namespace to SeriviceJig", "namespace", n.Namespace)
+	}
+
+	labels := map[string]string{"apps": name}
+	// Finally create service
+	svc := n.newServiceTemplate(name, n.Namespace, annotations, labels)
+	svc.Spec.Type = corev1.ServiceTypeNodePort
+	svc.Spec.Ports[0].NodePort = nodePort
```

Comment:
Is there a reason for having those setup after the creation of the svc and not directly inside the `newServiceTemplate` method, with the `nodePort` as an additional argument?

Reply:
you're right, at the moment it would be better to do as you suggested. I thought that keeping it generic will be beneficial as we will need also to expose services of type ClusterIp in a near future ;-)